File: partiql-plan/src/main/java/org/partiql/plan/rex/RexCase.java
package org.partiql.plan.rex;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.partiql.plan.Operand;
import org.partiql.plan.OperatorVisitor;

import java.util.List;

/**
 * Logical case (switch) expression abstract base class.
 */
public abstract class RexCase extends RexBase {

    /**
     * Creates a new case expression.
     * @param match the match expression, or {@code null} if none (operand 0)
     * @param branches the list of branches (not operands)
     * @param def the default expression, or {@code null} if none (not an operand)
     * @return the new case expression
     */
    @NotNull
    public static RexCase create(@Nullable Rex match, @NotNull List<Branch> branches, @Nullable Rex def) {
        return new Impl(match, branches, def);
    }

    /**
     * Creates a new branch.
     * @param condition the condition expression
     * @param result the result expression
     * @return the new branch
     */
    @NotNull
    public static Branch branch(@NotNull Rex condition, @NotNull Rex result) {
        return new Branch(condition, result);
    }

    /**
     * Gets the match expression, or {@code null} if none (operand 0).
     * @return the match expression, or {@code null} if none (operand 0)
     */
    @Nullable
    public abstract Rex getMatch();

    /**
     * Gets the list of branches (not operands).
     * @return the list of branches (not operands).
     */
    @NotNull
    public abstract List<Branch> getBranches();

    /**
     * Gets the default expression, or {@code null} if none (not an operand).
     * @return the default expression, or {@code null} if none (not an operand)
     */
    @Nullable
    public abstract Rex getDefault();

    @NotNull
    @Override
    protected final RexType type() {
        throw new UnsupportedOperationException("Derive type is not implemented");
    }

    @NotNull
    @Override
    protected List<Operand> operands() {
        Rex match = getMatch();
        if (match == null) {
            return List.of();
        }
        Operand c0 = Operand.single(match);
        return List.of(c0);
    }

    @Override
    public <R, C> R accept(OperatorVisitor<R, C> visitor, C ctx) {
        return visitor.visitCase(this, ctx);
    }

    /**
     * A branch of a case expression.
     */
    public static class Branch {

        @NotNull
        private final Rex condition;

        @NotNull
        private final Rex result;

        private Branch(@NotNull Rex condition, @NotNull Rex result) {
            this.condition = condition;
            this.result = result;
        }

        @NotNull
        public Rex getCondition() {
            return condition;
        }

        @NotNull
        public Rex getResult() {
            return result;
        }
    }

    private static class Impl extends RexCase {
        private final Rex match;
        private final List<Branch> branches;
        private final Rex def;

        private Impl(Rex match, List<Branch> branches, Rex def) {
            this.match = match;
            this.branches = branches;
            this.def = def;
        }

        @Nullable
        @Override
        public Rex getMatch() {
            return match;
        }

        @NotNull
        @Override
        public List<Branch> getBranches() {
            return branches;
        }

        @Nullable
        @Override
        public Rex getDefault() {
            return def;
        }
    }
}

File: partiql-plan/src/main/kotlin/org/partiql/plan/rex/RexCase.kt
package org.partiql.plan.rex

import org.partiql.types.PType

/**
 * Representative of the simple CASE-WHEN.
 */
public interface RexCase : Rex {

    public fun getMatch(): Rex?

    public fun getBranches(): List<Branch>

    public fun getDefault(): Rex?

    override fun <R, C> accept(visitor: RexVisitor<R, C>, ctx: C): R = visitor.visitCase(this, ctx)

    override fun getChildren(): Collection<Rex> {
        val children = mutableListOf<Rex>()
        val match = getMatch()
        val branches = getBranches()
        val default = getDefault()
        if (match != null) {
            children.add(match)
        }
        for (branch in branches) {
            children.add(branch.getCondition())
            children.add(branch.getResult())
        }
        if (default != null) {
            children.add(default)
        }
        return children
    }

    /**
     * TODO DOCUMENTATION
     */
    public interface Branch {
        public fun getCondition(): Rex
        public fun getResult(): Rex
    }
}

/**
 * Default [RexCase] meant for extension.
 */
internal class RexCaseImpl(match: Rex?, branches: List<RexCase.Branch>, default: Rex?) : RexCase {

    // DO NOT USE FINAL
    private var _match = match
    private var _branches = branches
    private var _default = default
    private var _children: Collection<Rex>? = null

    override fun getMatch(): Rex? = _match

    override fun getBranches(): List<RexCase.Branch> = _branches

    override fun getDefault(): Rex? = _default

    override fun getChildren(): Collection<Rex> {
        if (_children == null) {
            _children = super.getChildren()
        }
        return _children!!
    }

    override fun getType(): PType {
        TODO("Not yet implemented")
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is RexCase) return false
        if (_match != other.getMatch()) return false
        if (_branches != other.getBranches()) return false
        if (_default != other.getDefault()) return false
        return true
    }

    override fun hashCode(): Int {
        var result = _match.hashCode()
        result = 31 * result + _branches.hashCode()
        result = 31 * result + _default.hashCode()
        return result
    }

    /**
     * CASE-WHEN branch
     *
     * @param condition
     * @param result
     */
    internal class Branch(condition: Rex, result: Rex) : RexCase.Branch {

        // DO NOT USE FINAL
        private var _condition = condition
        private var _result = result

        override fun getCondition(): Rex = _condition

        override fun getResult(): Rex = _result
    }
}
